qiskit_ibm_provider/job/ibm_circuit_job.py

```diff
@@ -803,21 +816,33 @@ def _retrieve_result(self, refresh: bool = False) -> None:
 
         if not self._result or refresh:  # type: ignore[has-type]
             try:
-                result_response = self._api_client.job_result(
-                    self.job_id(), self._use_object_storage
-                )
-                self._set_result(result_response)
-                if self._status is JobStatus.ERROR:
-                    # Look for error message in result response.
-                    self._check_for_error_message(result_response)
+                api_result = self._runtime_client.job_results(self.job_id())
+                self._set_result(api_result)
+                # TODO: Look for error message in result response.
+                # if self._status is JobStatus.ERROR:
+                #     self._check_for_error_message(api_result)
             except ApiError as err:
                 if self._status not in (JobStatus.ERROR, JobStatus.CANCELLED):
                     raise IBMJobApiError(
                         "Unable to retrieve result for "
                         "job {}: {}".format(self.job_id(), str(err))
                     ) from err
 
-    def _set_result(self, raw_data: Optional[Dict]) -> None:
+    def _check_result_for_errors(self, raw_data: str) -> None:
+        """Checks whether the job result contains errors
+
+        Args:
+            raw_data: Raw result data.
+
+        Raises:
+            IBMJobError: If an error was found in the result
+
+        """
+        result = re.search("JobError: '(.*)'", raw_data)
+        if result is not None:
+            raise IBMJobError(result.group(1))
```

Comment:
Looking for `JobError` in the result is really not the best approach, since the job can fail without any data returned. It's best to verify the job has indeed failed. See similar code [here](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/runtime_job.py#L369-L374).

Reply:
Done. Now everything is checked via `self._status == JobStatus.ERROR` and if an error was detected, `error_message()` is called. To enable informative error descriptions, I still parse for `JobError` and return the result if found, but that's not how we decide the job failed.

I also attempt to use the qiskit-runtime `reason` mechanism, but right now it's not informative enough (just returns "Error").